feat(subscription): add mixed cart documentation (#1968)
shopware/docs: products/extensions/subscriptions/guides/mixed-checkout.md

---
nav:
  title: Mixed checkout
  position: 20

---
# Mixed subscription checkout

::: info
Available since Shopware version 6.7.4.0
:::

This guide describes how the mixed carts in the subscription feature works and how extensions should integrate with it.

> Mixed carts let customers buy subscription products and one‑time products in the same checkout while keeping subscription calculation isolated and predictable.

Reviewer: ```suggestion
Mixed carts let customers buy subscription products and one‑time products during a single checkout while keeping subscription calculation isolated and predictable.
```

Please familiarise yourself with the [concept](../concept.md) first before continuing here.

> 

Reviewer: I'd love a short paragraph here or after the overview, why mixed carts are such an issue with Shopware and what difficulties we had to overcome to make it work with rules etc, while being performant and not rewriting the whole cart logic. 

I think it could help 3rd party devs to understand, why we use managed carts.

## Overview

Subscription line items are ordinary product line items in the main cart, but they carry subscription plan and interval IDs in their payload.
During cart calculation, line items containing subscription plan and interval IDs in their payload are collected and grouped by plan and interval.
For each group a derived _managed_ subscription context and a _managed_ subscription cart are created and calculated using the subscription cart calculation path.
These represent the context and content of the later generated subscription.

Managed subscription contexts and carts are persisted in the database as well.
They are linked back to the main context by the `subscription_cart` database table.

> 

Reviewer: A little bonus, if you're intersted: 
A real cherry on top could be an illustration here: 

Shopware (Parent) Context 1:n Subscription Managed Context with showing the extension Map and such
Maybe ask the docs team or come up with something on your own.

## Retrieving information

You can access the managed carts through the cart extension named `subscriptionManagedCarts`, which maps keys in the form `<plan-id>-<interval-id>` to their corresponding [managed cart](../concept.md#subscription-cart).
The sales channel context extension named `subscriptionManagedContexts` provides the same mapping for [managed sales channel contexts](../concept.md#subscription-context).

> The mapping is considered stable and the plan and interval IDs can be retrieved by splitting the composite ID.

Reviewer: What do you mean with `stable` ?
It could be misleading, as `stable` usually means, that the structure is considered productive. I think you want to say sth else here?

Author: I would like to say that this composite ID will not change, in terms of either the order or the fact that plan and interval IDs are used. Splitting the ID is the main way of obtaining both IDs, so there is no need to worry that this process will not be successful in the future

When an order is placed from a mixed cart, the order will contain an `initialSubscriptions` extension that includes all created subscriptions.
As any subsequent orders are generated per subscription, the orders will contain a `subscriptionId` / `subscription` extension instead.

## Manipulate mixed cart

> Subscription mixed carts does not change how the main cart is manipulated.

Reviewer: ```suggestion
Subscription mixed carts do not interfere with the main cart calculation.
```

Do you mean this?

Author: No, I mean: The main cart can be manipulated like you (as developer) are used to, in contrast to the separate checkout where you need to have a subscription collector / processor to change the cart.
